File: src/IsbInterop/Scope.vb
Imports IsbInterop.Accessory
Imports IsbInterop.Accessory.Proxies

''' <summary>
''' Область видимости.
''' </summary>
Friend Class Scope
  Implements IScope
  ''' <summary>
  ''' Рабочий набор.
  ''' </summary>
  Private ReadOnly _workingSet As New Stack(Of IBaseIsbObject)()

  ''' <summary>
  ''' Приложение IS-Builder.
  ''' </summary>
  Public ReadOnly Property Application As IApplication Implements IScope.Application

#Region "IDisposable"

  ''' <summary>
  ''' Выполняет очистку.
  ''' </summary>
  Public Sub Dispose() Implements IDisposable.Dispose
    While _workingSet.Count > 0
      Dim requisite = _workingSet.Pop()
      Try
        requisite.Dispose()
        ' Гасим исключение, если объект уже освобожден.
      Catch exception As ObjectDisposedException
      End Try
    End While

    _workingSet.Clear()
  End Sub

#End Region

  ''' <summary>
  ''' Добавляет объект в область видимости.
  ''' </summary>
  ''' <param name="isbObject">Объект IS-Builder.</param>
  Public Sub Add(isbObject As IBaseIsbObject)
    _workingSet.Push(isbObject)
  End Sub

  ''' <summary>
  ''' Конструктор.
  ''' </summary>
  ''' <param name="rcwApp">RCW-объект IApplication.</param>
  Public Sub New(rcwApp As Object)
    Application = New Application(rcwApp, Me)
  End Sub
End Class

''' <summary>
''' Интерфейс области видимости.
''' </summary>
Public Interface IScope
  Inherits IDisposable
  ''' <summary>
  ''' Приложение IS-Builder.
  ''' </summary>
  ReadOnly Property Application As IApplication
End Interface

File: src/IsbInterop/Scope.cs
﻿using IsbInterop.Accessory;
using IsbInterop.Accessory.Proxies;
using System;
using System.Collections.Generic;
using IsbInterop.System;
using IsbInterop.System.Proxies;

namespace IsbInterop
{
  /// <summary>
  /// Область видимости.
  /// </summary>
  internal class Scope : IScope
  {
    /// <summary>
    /// Рабочий набор.
    /// </summary>
    private readonly Stack<IBaseIsbObject> _workingSet = new Stack<IBaseIsbObject>();

    /// <summary>
    /// Приложение IS-Builder.
    /// </summary>
    public IApplication Application { get; }

    #region IDisposable

    /// <summary>
    /// Выполняет очистку.
    /// </summary>
    public void Dispose()
    {
      while (_workingSet.Count > 0)
      {
        var requisite = _workingSet.Pop();
        try
        {
          requisite.Dispose();
        }
        catch (ObjectDisposedException)
        {
          // Гасим исключение, если объект уже освобожден.
        }
      }

      _workingSet.Clear();
    }

    #endregion

    /// <summary>
    /// Добавляет объект в область видимости.
    /// </summary>
    /// <param name="isbObject">Объект IS-Builder.</param>
    public void Add(IBaseIsbObject isbObject)
    {
      _workingSet.Push(isbObject);
    }

    /// <summary>
    /// Конструктор.
    /// </summary>
    /// <param name="rcwApp">RCW-объект IApplication.</param>
    public Scope(object rcwApp)
    {
      Application = new Application(rcwApp, this);
    }
  }

  /// <summary>
  /// Интерфейс области видимости.
  /// </summary>
  public interface IScope : IDisposable
  {
    /// <summary>
    /// Приложение IS-Builder.
    /// </summary>
    IApplication Application { get; }
  }
}
